### ISSUE TITLE
Accessibility conformance compile-time checks

### ISSUE THREAD

**Issue by twerske** [angular/angular#42897]
### Which @angular/* package(s) are relevant/releated to the feature request?

compiler-cli, compiler

### Description

Introduce compile-time conformance checks to guard against accessibility anti-patterns (similar to https://github.com/angular/angular/issues/42690).

### Proposed solution

The Angular compiler can perform extra checks on the user application to guard against common pitfalls. We already have the available parsing and reporting infrastructure, and there's work happening in the direction of expanding the number of compile-time checks.

A few possible candidates for helpful a11y conformance rules are:
- Check for alt text on templates and that alt text is not repetitive (images with "image" in the text)
- Page title defined in the TitleService is unique to each route (or not empty).
- Check that ARIA attributes, props and respective values are valid (one example is the known values for "role" as described in https://github.com/angular/angular/issues/754)
- Ensures that the heading, anchor and button elements have content in it
- Forms should have labels and associated controls

### Alternatives considered

An alternative is to delegate these checks to linters, such as eslint or codelyzer, which have some coverage for these rules. Even though this would reduce maintenance, we'd also be limited to only applications using the linting tooling and have explicitly enabled these checks.

**Comment by Airblader** [comment]
> we'd also be limited to only applications using the linting tooling and have explicitly enabled these checks

So why not just ship it by default in Angular CLI instead of re-implementing a custom solution in the framework?

**Comment by Airblader** [comment]
Unrelated side note: so many feature requests have been closed due to insufficient votes saying "just implement it as a library", when in most cases that isn't possible. Then this issue comes along, something that not only can but already has been implemented in various external tools, but it gets to entirely skip the voting process.

**Comment by Airblader** [comment]
And another comment, but this time related: won't this be confusing / annoying for beginners who are just exploring web development and Angular? They'd see a bunch of warnings / errors (?) for things they don't understand and that aren't actually crucial for the framework to work.

**Comment by Airblader** [comment]
There's also all kinds of valid scenarios that will make it very much non-trivial to detect these things reliably at compile-time:

```
<!-- Attributes like title could be set through directives -->
<img sec="..." appImageTitle="I will be the title" />

<!-- Does this button have content or not? -->
<button libMenuButton></button>
```

```
@HostBinding("role")
public get role() {
    return this.hadOnInit ? "invalidrole" : "textbox";
}
```

```
// Good luck validating that this is unique or non-empty :-)
this.backend.fetchPageTitle().subscribe(title => this.titleService.setTitle(title));
```

**Comment by flash-me** [comment]
First of all, this would be really awesome.

Which a11y rules will be supportet and checked against? The WCAG defines the specs and what is possible or not. Like which roles exist, which aria attributes are allowed etc.

But each country comes with different rules and conformances

-  In America there is the ADA ( https://www.adachecklist.org/ ) 
-  In Germany there is BITV (https://www.bitvtest.de/bitv_test.html) 
- etc...

**To point out the complexity of this topic:**

> Ensures that the heading, anchor and button elements have content in it

What about Buttons that only have icons in it? For those we use `aria-label` to describe the button, but it doesn't have an actual content. Would this lead to an error?....


> And another comment, but this time related: won't this be confusing / annoying for beginners who are just exploring web development and Angular? They'd see a bunch of warnings / errors (?) for things they don't understand and that aren't actually necessary for the application to work.

You can't be serious, or? Aria Attributes and roles ARE part of web development. The www would be much better place, if people would have learned those topics right away. A lot of time and effort has been put in the specs to enable providing semantical meaning. It's totally rude to even tell people to ignore this kind of topic. Technology should be used to remove barriers and provide everyone with the same possibilites without restrictions.

cheers
flash ⚡

**Comment by Airblader** [comment]
I don't disagree that a11y is important, but expecting someone to learn all of that before they ever get to write any code isn't reasonable. If you're a beginner who barely understands JS, you're already struggling to learn TS and Angular concepts. Throwing in the expectation to also learn a11y in the same step because otherwise Angular yells at you is a serious hurdle.

**Comment by flash-me** [comment]
Well, then you might learn longer.
Not surprising that people think "let's watch youtube videos for 3 months and start a developer career". 
Thank you for giving a f@ck about the important things in life....

cheers
flash ⚡

**Comment by flash-me** [comment]
Don't get me wrong @Airblader . I do insist that people should include this kind of topics of the very beginning. And I also do understand your point of view. What I think is the web developer community needs a huge mindset change about a11y.

But like many other options in the toolchain this could be another flag to enable / disable the checks without enforcing it.

cheers
flash ⚡

**Comment by Airblader** [comment]
Please familiarize yourself with the Code of Conduct. Thanks. I refrain from further discussing this with you.

**Comment by flash-me** [comment]
> Edit: this referred to a now edited out comment.

Yes and I do apologize for my wording. You are right. I'm just kinda "sensible" about this topic. I'm working hand in hand with dudes that need to use screen reader. But this shouldn't be an excuse to lose respect.

cheers
flash ⚡

**Comment by pkozlowski-opensource** [comment]
Relevant / similar issue opened in the past #754

**Comment by jnizet** [comment]
I would add to the objections of @Airblader that many times, the Angular compiler wouldn't even be able to make the appropriate checks since it doesn't have the rendered page at its disposal, but only the template of a component. So, for example

- how would it check the title of the pages if it's dynamic/i18ned?
- how would it check if a role is correct if it's dynamic?
- how would it check if an input has a label if the two input and the label are in separate components
- how would it check if an input has a label if the IDs are dynamically generated (which is often the case in dynamic forms)?

An E2E test is the best place, IMHO, to put such checks, and plugins already exist to do that (like https://github.com/component-driven/cypress-axe for example).
